### PR TITLE
Tests: Functional tests for GPP /cookie_sync endpoint.

### PR DESCRIPTION
**Pull request by osulzhenko**

### CODE REVIEW

File: src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        def metric = this.prebidServerService.sendCollectedMetricsRequest()
+        assert metric["cookie_sync.generic.tcf.blocked"] == 1
+    }
+
+    def "PBS cookie sync request should set GBPR to 1 when GPP contains 2"() {
+        given: "Request without GDPR and invalid GPP"
+
+        def gppSid = "${PBSUtils.randomNumber},${PBSUtils.randomNumber}, 2"
+        def cookieSyncRequest = CookieSyncRequest.defaultCookieSyncRequest.tap {
+            it.bidders = []
```
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy:68]: Do you need to explicitly override seemingly irrelevant parameter?

File: src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        then: "Response should contain gdpr as 1"
+        assert response.gdpr == 1
+    }
+
+    def "PBS cookie sync request should set GBPR to 0 when GPP contains not 2"() {
+        given: "Request without GDPR and invalid GPP"
+
+        def gppSid = "0, 1"
+        def cookieSyncRequest = CookieSyncRequest.defaultCookieSyncRequest.tap {
+            it.bidders = []
```
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy:86]: Do you need to explicitly override seemingly irrelevant parameter?

File: src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            it.gpp = null
+            it.gppSid = gppSid
+            it.gdpr = null
+        }
+
+        when: "PBS processes cookie sync request"
+        def response = prebidServerService.sendCookieSyncRequest(cookieSyncRequest)
+        then: "Response should contain gdpr as 0"
```
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy:94]: Please add a blank line between sections.

File: src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        assert response.gdpr == 0
+    }
+
+    def "PBS cookie sync request should reflect warning when GPP is contains 2  and gdpr is not 1"() {
```
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/tests/GppCookieSyncSpec.groovy:98]: 1) "reflect" -> "respond with"
2) "PBS cookie sync request should respond with a warning when GPP SID contains 2 and gdpr is not 1"

**Review by Net-burst — CHANGES_REQUESTED** [review]
Very good overall, but left some remarks. Most of them apply to all of the tests.

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+            it.gpp = null
+            it.gppSid = gppSid
+            it.gdpr = 1
+        }
+
+        when: "PBS processes cookie sync request"
+        def response = prebidServerService.sendCookieSyncRequest(cookieSyncRequest)
+
+        then: "Response should contain warning"
+        assert response.ext.warnings == "GPP scope does not match TCF2 scope"
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:127]: Same

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+        and: "Metric should not contain cookie_sync.FAMILY.tcf.blocked"
+        assert prebidServerService.sendCollectedMetricsRequest() == [ "cookie_sync.generic.tcf.blocked": 0 ]
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:130]: Redundant

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    def "PBS cookie sync request should respond with warning when GPP string is invalid"() {
+        given: "Request with invalid GPP"
+        def invalidGpp = PBSUtils.getRandomString(20)
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:135]: just `randomString`

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+        when: "PBS processes cookie sync request"
+        def response = prebidServerService.sendCookieSyncRequest(cookieSyncRequest)
+
+        then:
+        expect: "Response should contain warning"
+        assert response.ext.warnings == "GPP string invalid: ERROR MESSAGE FROM GPP LIBRARY"
+
+        and: "Metric should not contain cookie_sync.FAMILY.tcf.blocked"
+        assert prebidServerService.sendCollectedMetricsRequest() == [ "cookie_sync.generic.tcf.blocked": 0 ]
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:148]: Same. Redundant

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        def invalidGpp = PBSUtils.getRandomString(20)
+        def cookieSyncRequest = CookieSyncRequest.defaultCookieSyncRequest.tap {
+            it.gpp = invalidGpp
+        }
+
+        when: "PBS processes cookie sync request"
+        def response = prebidServerService.sendCookieSyncRequest(cookieSyncRequest)
+
+        then:
+        expect: "Response should contain warning"
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:144]: use `then`, `expect` keyword doesn't present

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        assert response.ext.warnings == "GPP string invalid: ERROR MESSAGE FROM GPP LIBRARY"
+
+        and: "Metric should not contain cookie_sync.FAMILY.tcf.blocked"
+        assert prebidServerService.sendCollectedMetricsRequest() == [ "cookie_sync.generic.tcf.blocked": 0 ]
+    }
+
+    def "PBS should copy regs.gpp to user.consent when gppSid contains 2, gpp is TCF2-EU and user.consent isn't specified without any warning from cookie sync request"() {
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:151]: We can't test this scenario

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        def cookieSyncResponse = prebidServerService.sendCookieSyncRequest(cookieSyncRequest, uidsCookie)
+
+        then: "CookieSync response should contain warning"
+        assert cookieSyncResponse.ext.warnings == "GPP TCF2 string does not match gdpr_consent"
+
+        and: "Metric should not contain cookie_sync.FAMILY.tcf.blocked"
+        assert prebidServerService.sendCollectedMetricsRequest() == ["cookie_sync.generic.tcf.blocked": 0]
+    }
+
+    def "PBS should copy gpp to usPrivacy when gppSid contains 6, gpp is TCF2-EU and us_privacy isn't specified without warnings"() {
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:218]: Same we can't test it

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+        assert bidderRequest.regs.gpp == gppConsent as String
+
+        and: "Response should NOT contain warning"
+        assert cookieSyncResponse.ext.warnings == null
+
+        and: "Metric should not contain cookie_sync.FAMILY.tcf.blocked"
+        assert prebidServerService.sendCollectedMetricsRequest() == ["cookie_sync.generic.tcf.blocked": 0]
+    }
+
+    def "PBS should copy regs.gpp to user.consent when gppSid contains 2, gpp is TCF2-EU and user.consent are different with warning from cookie sync request"() {
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:188]: Same. We can't test it

File: src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+        def uidsCookie = UidsCookie.defaultUidsCookie
+
+        when: "PBS processes cookie sync request"
+        def cookieSyncResponse = prebidServerService.sendCookieSyncRequest(cookieSyncRequest, uidsCookie)
+
+        then: "CookieSync response should contain warning"
+        assert cookieSyncResponse.ext.warnings == "GPP TCF2 string does not match gdpr_consent"
+
+        and: "Metric should not contain cookie_sync.FAMILY.tcf.blocked"
+        assert prebidServerService.sendCollectedMetricsRequest() == ["cookie_sync.generic.tcf.blocked": 0]
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/privacy/GppCookieSyncSpec.groovy:255]: Redundant

**Review by marki1an — CHANGES_REQUESTED** [review]

**Review by Net-burst — APPROVED** [review]
LGTM

**Pull request merged**